### DISCUSSION TITLE
Recovering intrinsic camera matrix from simulation camera

### DISCUSSION THREAD

**Discussion by sunshineinsandiego** [surgical-robotics-ai/surgical_robotics_challenge#84] [Q&A]
Hi - I am trying to recover an homogenous intrinsic camera matrix of the form:

```math
\begin{bmatrix}
f_x & 0 & c_x & 0\\
0 & f_y & c_y & 0\\
0 & 0 & 1 & 0
\end{bmatrix}
```

From the simulator camera specifications [here](https://github.com/collaborative-robotics/surgical_robotics_challenge/blob/master/ADF/world/world_stereo.yaml#L42). @adnanmunawar noted [here](https://github.com/collaborative-robotics/surgical_robotics_challenge/discussions/75#discussioncomment-5864226) that I should be able to recover the intrinsic matrix given the near plane, far plane, FOV, and aspect ratio, and I would like to confirm that I am doing this correctly. 

Using this [source](https://stackoverflow.com/questions/27699114/intrinsic-camera-parameters-to-opengl-perspective-matrix-near-far-parameters) and this [source](http://www.terathon.com/gdc07_lengyel.pdf), it appears that the conversion from simulation camera parameters to OpenGL projection matrix is: 

```math
\begin{bmatrix}
e & 0 & 0 & 0\\
0 & e/a & 0 & 0\\
0 & 0 & -(f+n)/(f-n) & -2fn/(f-n)\\
0 & 0 & -1 & 0
\end{bmatrix}
```

Any advice on how to go from this projection matrix to the camera intrinsic matrix? [This](https://github.com/facebookresearch/pytorch3d/issues/1014) seems like another possible step, but not sure.

Thanks for your help!

**Comment by adnanmunawar** [comment]
Hello, have you taken a look at this [document](https://github.com/collaborative-robotics/surgical_robotics_challenge/blob/master/docs/camera_conventions.md) by @jabarragann ?

**Reply by sunshineinsandiego** [reply]
Ah yes! I have, but when I test it by constructing the simulator frustum and the opencv camera intrinsic matrix, then projecting an identical point in 3D space onto the image plane, I get different answers.

But if that conversion in the doc is correct, perhaps the error in projection is due to something other than the mapping between the frustum vs. K matrix?

**Reply by adnanmunawar** [reply]
Hmm, it is possible. Let's see what @jabarragann thinks.

**Reply by jabarragann** [reply]
Hi @sunshineinsandiego,

Apologies for the late reply. In this [document](https://github.com/surgical-robotics-ai/surgical_robotics_challenge/blob/master/docs/camera_conventions.md), we projected some 3D points specified in the needle reference frame and projected them back into the image using opencv, so I am confident that the formulas presented here are correct. If you describe your setup and the problems you are facing, I am more than happy to help you debug the problem.

**Reply by sunshineinsandiego** [reply]
Thanks @jabarragann. Sounds like you ran the tests that I'm trying to tie out. I think the issue that I was having was related to the external camera matrix parameters I was setting. Through a bit of trial and error it seems to be fixed now, but I thank you for your offer to help! Will close this issue.

**Reply by jabarragann** [reply]
Fantastic, I am glad it got fixed!
